File: 323_Number_of_Connected_Components_in_an_Undirected_Graph.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
// DFS time complexity: O(n)
private:
    void dfs(vector<vector<int>>& graph, vector<bool>& seen, int s){
        if (seen[s]) return;
        seen[s] = true;
        for (int i : graph[s]){
            dfs(graph, seen, i);
        }
    }
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> graph(n, vector<int>());
        for (auto edge : edges){
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        int cnt = 0;
        vector<bool> seen(n, false);
        for (int i = 0; i < n; ++i){
            if (seen[i]) continue;
            dfs(graph, seen, i);
            cnt++;
        }
        return cnt;
    }
};

int main(){
    Solution* solution = new Solution();
    vector<vector<int>> edges{{0, 1}, {1, 2}, {3, 4}};
    cout << solution->countComponents(5, edges) << "\n";
    return 0;
}

File: 323_Number_of_Connected_Components_in_an_Undirected_Graph.py
class Solution():
    def countComponents(self, n, edges):
        # BFS time complexity: O(n)
        def bfs(s):
            queue = [s]
            while queue:
                t = queue.pop()
                seen[t] = True
                for n in graph[t]:
                    if not seen[n]:
                        queue.append(n)

        graph = [set() for _ in range(n)]
        for e1, e2 in edges:
            graph[e1].add(e2)
            graph[e2].add(e1)

        cnt = 0
        seen = [False for _ in range(n)]
        for i in range(n):
            if seen[i]:
                continue
            bfs(i)
            cnt += 1
        return cnt


solution = Solution()
res = solution.countComponents(5, [[0, 1], [1, 2], [3, 4]])
print("Result: ", res)

res = solution.countComponents(5, [[0, 1], [1, 2], [2, 3], [3, 4]])
print("Result: ", res)